File: data_structures_and_algorithms/Code_401/class-05/solutions/java/LinkedList.java
import java.util.*;

class LinkedList <T> {
  public static void main(String[] args){
    LinkedList<Integer> ints = new LinkedList<>();
    ints.insert(1);
    ints.insert(5);
    ints.insert(6);
    System.out.println(ints);
    System.out.println(ints.includes(5));
    System.out.println(ints.includes(45));
  }
  Node<T> head;

  public void insert(T value){
    this.head = new Node<T>(value, this.head);
  }

  public boolean includes(T value){
    Node<T> current = head;
    while(current != null){
      if(current.value.equals(value)) return true;
      current = current.next;
    }
    return false;
  }

  public String toString(){
    StringBuilder stringified = new StringBuilder();
    Node<T> current = this.head;

    while(current != null){
      stringified.append("{" + current.value + "} -> ");
      current = current.next;
    }
    stringified.append("NULL");
    return stringified.toString();
  }

}

class Node<T> {
  T value;
  Node<T> next;

  public Node(T value, Node<T> next){
    this.value = value;
    this.next = next;
  }
}

File: data_structures_and_algorithms/Code_401/class-05/solutions/python/linked_list.py
class LinkedList:
    """
    Put docstring here
    """

    def __init__(self, values=None):
        self.head = None
        if values:
            for value in reversed(values):
                self.insert(value)

    def __str__(self):
        text = ""

        current = self.head

        while current:
            text += f"{{ {current.value} }} -> "
            current = current.next

        return text + "NULL"

    def insert(self, value):
        self.head = Node(value, self.head)

    def includes(self, value):
        current = self.head
        while current:
            if current.value == value:
                return True

            current = current.next

        return False

    def append(self, value):

        node = Node(value)

        if not self.head:
            self.head = node
            return

        current = self.head

        while current.next:
            current = current.next

        if current:
            current.next = node

    def insert_before(self, before_value, insert_value):

        if not self.head:
            raise TargetError

        if self.head.value == before_value:
            self.head = Node(insert_value, self.head)
            return

        current = self.head

        while current.next:
            if current.next.value == before_value:
                current.next = Node(insert_value, current.next)
                return
            current = current.next

        raise TargetError

    def insert_after(self, after_value, insert_value):

        current = self.head

        while current:
            if current.value == after_value:
                current.next = Node(insert_value, current.next)
                return

            current = current.next

        raise TargetError

    def kth_from_end(self, k):
        current = self.head
        length = 0
        while current:
            length += 1
            current = current.next

        current = self.head

        steps_remaining = length - k - 1

        if steps_remaining < 0:
            raise TargetError

        for _ in range(steps_remaining):
            current = current.next
            if not current:
                raise TargetError

        return current.value


class Node:
    def __init__(self, value, next_=None):
        self.value = value
        self.next = next_


class TargetError(Exception):
    pass
